### ISSUE TITLE
Validate is a flat line for values unseen y

### ISSUE THREAD

**Issue by bpmforge** [jinglescode/time-series-forecasting-tensorflowjs#5]
Every time I run any test I am getting a flat line. so when I train on 80% data and validate on the 20% remaining, line [225](https://github.com/jinglescode/time-series-forecasting-tensorflowjs/blob/master/index.js#L225) gives results:
`(220) [51.777862548828125, 51.78431701660156, 51.84973907470703, 51.965545654296875, 52.03998947143555, 52.07977294921875, 52.12873458862305, 52.22467041015625, 52.245277404785156, 52.23814392089844, 52.30770492553711, 52.34142303466797, 52.36872100830078, 52.380374908447266, 52.41004180908203, 52.414215087890625, 52.41645431518555, 52.433067321777344, 52.436336517333984, 52.4455451965332, 52.45906066894531, 52.465614318847656, 52.461029052734375, 52.469947814941406, 52.47417449951172, 52.47554397583008, 52.48143005371094, 52.48564147949219, 52.48596954345703, 52.49354934692383, 52.50459671020508, 52.51314163208008, 52.52178192138672, 52.5322265625, 52.54399108886719, 52.551612854003906, 52.55399703979492, 52.56175231933594, 52.57003402709961, 52.564884185791016, 52.56238555908203, 52.568458557128906, 52.57207489013672, 52.576416015625, 52.585086822509766, 52.589359283447266, 52.58890151977539, 52.58769989013672, 52.589717864990234, 52.59192657470703, 52.597312927246094, 52.5996208190918, 52.60268783569336, 52.606224060058594, 52.60929489135742, 52.61307144165039, 52.622291564941406, 52.628456115722656, 52.63402557373047, 52.638301849365234, 52.640830993652344, 52.640846252441406, 52.640708923339844, 52.642784118652344, 52.6439208984375, 52.64525604248047, 52.646976470947266, 52.64796447753906, 52.649070739746094, 52.651641845703125, 52.65275955200195, 52.653404235839844, 52.654991149902344, 52.65419387817383, 52.65365982055664, 52.655967712402344, 52.656612396240234, 52.656105041503906, 52.65705108642578, 52.65494918823242, 52.653289794921875, 52.65436553955078, 52.655303955078125, 52.65666580200195, 52.65773010253906, 52.65778350830078, 52.658077239990234, 52.659000396728516, 52.65916442871094, 52.65989303588867, 52.66079330444336, 52.66077423095703, 52.660621643066406, 52.66080856323242, 52.660499572753906, 52.66173553466797, 52.6634407043457, 52.6645622253418, 52.66503143310547, 52.66551971435547, …]`

**Comment by jinglescode** [comment]
Hi @feuerstarta, 

Do check out [Why isn't my Model Performing?](https://github.com/jinglescode/time-series-forecasting-tensorflowjs#why-isnt-my-model-performing).

Because the market dipped and recovered within a month or two, this has never happened in history. Since you are using 80/20, the model does not have the chance to learn that pattern. 

Also, this is a very simple model, feel free to tweak the model, for better performance. Try building more features on the data as well, possibly adding "difference from previous day".

**Comment by bpmforge** [comment]
See that still makes no sense to be a flat line. the flat line means it is not taking any of the input of the unseen_x into the account. If you do a full predict on the 0-100% of the 80/20 learn, it predicts value till the 80% mark then just flatlines from the last predicted number. If you keep that 100% prediction and do 90/10 split it will go to the 90% mark then flatlines at the 90% mark with the last predicted value. This is showing that the model works only for the section it trains and can not do a prediction period as the model only reports the last recorded train.

**Comment by jinglescode** [comment]
This is the challenge of machine learning, that a model, out of the box, doesn't work for all kinds of data. There is a lot of parameter tuning involve. There may need to do more feature engineering. Or there is a need to change the model architecture. 

Doing a quick search, you can see that you are not the only one who faced this problem:

- [LSTM always predict same values for all inputs in the batch](https://discuss.pytorch.org/t/lstm-always-predict-same-values-for-all-inputs-in-the-batch/53082)
- [Outputs of LSTM Model all Very Similar or the Same](https://discuss.pytorch.org/t/outputs-of-lstm-model-all-very-similar-or-the-same/84370)
- [Pytorch RNN always gives the same output for multivariate time series](https://www.reddit.com/r/learnmachinelearning/comments/9zugw8/pytorch_rnn_always_gives_the_same_output_for/)

I do not have an answer for you, unfortunately. Some suggestions would be:
- try tweaking the normalize the data portion (`.div(tf.scalar(10)` might be causing problems [line 19 and 20](https://github.com/jinglescode/time-series-forecasting-tensorflowjs/blob/master/model.js))
- include more data features, maybe can include volume as well
- include more data features, maybe trading indicators such as MACD, RSI, Bollinger bands
- build another model architecture

I'm working on a Python version.
